app/database/build_store.py

```diff
@@ -0,0 +1,439 @@
+from collections import OrderedDict
+import json
+from queue import Queue
+from typing import List, Optional
+from sqlalchemy import func
+
+from app.database.connection import Connection
+from app.database.schema import AtomsSchema, BuildSchema, FailedArtifactDirectoriesSchema, FailedSubjobAtomPairsSchema, SubjobsSchema
+from app.common.build_artifact import BuildArtifact
+from app.master.atom import Atom
+from app.master.build import Build
+from app.master.build_request import BuildRequest
+from app.master.build_fsm import BuildState
+from app.master.subjob import Subjob
+from app.util.exceptions import ItemNotFoundError
+from app.util.log import get_logger
+
+
+# pylint: disable=protected-access
+class BuildStore:
+    """
+    Build storage service that stores and handles all builds.
+    """
+    _logger = get_logger(__name__)
+    _cached_builds_by_id = OrderedDict()
+
+    @classmethod
+    def get(cls, build_id: int, allow_incompleted_builds=False) -> Optional[Build]:
+        """
+        Returns a build by id.
+        :param build_id: The id for the build whose status we are getting
+        """
+        build = cls._cached_builds_by_id.get(build_id)
+        if build is None:
+            cls._logger.info('Requested build (id: {}) was not found in cache, checking database.'.format(build_id))
+            build = cls._reconstruct_build(build_id, allow_incompleted_builds=allow_incompleted_builds)
+            if build is not None:
+                cls._cached_builds_by_id[build_id] = build
+                cls._logger.notice('Build (id: {}) was added to cache.'.format(build_id))
+
+        return build
+
+    @classmethod
+    def get_range(cls, start: int, end: int, allow_incompleted_builds=False) -> List[Build]:
+        """
+        Returns a list of all builds.
+        :param start: The starting index of the requested build.
+        :param end: 1 + the index of the last requested element, although if this is greater than the total number
+                    of builds available the length of the returned list may be smaller than (end - start).
+        """
+        builds = []
+        # Add 1 to start & end because we're create build_id's, not indices
+        for build_id in range(start + 1, end + 1):
+            try:
+                builds.append(cls.get(build_id, allow_incompleted_builds=allow_incompleted_builds))
```

Comment:
Sqlalchemy lets you define a schema such that queries for related tables are automatically loaded without needing to query explicitly. I kind of hinted at it in other comments, but the relevant docs are:
http://docs.sqlalchemy.org/en/latest/orm/basic_relationships.html
http://docs.sqlalchemy.org/en/latest/orm/relationship_api.html#sqlalchemy.orm.relationship.params.lazy

Using relationships would allow you to only do an explicit query for the BuildSchema object, and the related queries for subjobs/atoms would happen automatically (immediately or as needed, depending on the `lazy` setting).